src/Functions/multiIf.cpp

```diff
@@ -239,6 +260,68 @@ class FunctionMultiIf final : public FunctionIfBase
 
         return res;
     }
+
+private:
+    static void executeShortCircuitArguments(ColumnsWithTypeAndName & arguments)
+    {
+        int last_short_circuit_argument_index = checkShirtCircuitArguments(arguments);
+        if (last_short_circuit_argument_index < 0)
+            return;
+
+        /// In multiIf we should execute the next condition only
+        /// if all previous once are false. So, we will filter
+        /// the next condition by inverted disjunction of previous once.
+        /// The next expression should be executed only if it's condition is
+        /// true and all previous conditions are false. So, we will
+        /// use default_value_in_expanding = 0 while extracting mask from
+        /// executed condition and filter expression by this mask.
+
+        IColumn::Filter current_mask;
+        MaskInfo current_mask_info = {.has_once = true, .has_zeros = false};
+        IColumn::Filter mask_disjunctions = IColumn::Filter(arguments[0].column->size(), 0);
+        MaskInfo disjunctions_mask_info = {.has_once = false, .has_zeros = true};
+
+        int i = 1;
+        while (i <= last_short_circuit_argument_index)
+        {
+            auto & cond_column = arguments[i - 1].column;
+            /// If condition is const or null and value is false, we can skip execution of expression after this condition.
+            if ((isColumnConst(*cond_column) || cond_column->onlyNull()) && !cond_column->empty() && !cond_column->getBool(0))
+            {
+                current_mask_info.has_once = false;
+                current_mask_info.has_zeros = true;
+            }
+            else
+            {
+                current_mask_info = getMaskFromColumn(arguments[i - 1].column, current_mask, false, &mask_disjunctions, 0, true);
+                maskedExecute(arguments[i], current_mask, current_mask_info, false);
+            }
+
+            /// Check if the condition is always true and we don't need to execute the rest arguments.
+            if (!current_mask_info.has_zeros)
+                break;
+
+            ++i;
+            if (i > last_short_circuit_argument_index)
+                break;
+
+            /// Make a disjunction only if it make sense.
+            if (current_mask_info.has_once)
+                disjunctions_mask_info = disjunctionMasks(mask_disjunctions, current_mask);
```

Comment:
As well as we use `mask_disjunctions` in `getMaskFromColumn` above, do we really need do disjoin masks here?
I feel like we can do it inside `getMaskFromColumn`. Not like the code will be easier ...

Reply:
We can do it inside `getMaskFromColumn` and make this function more complicated)
Now I think that there is a good idea to split `getMaskFromColumn`  into several functions with different additions. It will produce more code and maybe some code duplication, but can make it a bit easier to read. What do you think?